documentation/NETFramework-NGEN.md

```diff
@@ -0,0 +1,118 @@
+of `MSBuild.exe` and 2) Not all resolvers actually live under this directory; the system allows them to be placed anywhere.
+It is unfortunately also not straightforward to add a binding redirect with a [`codeBase`](https://learn.microsoft.com/en-us/dotnet/framework/configure-apps/file-schema/runtime/codebase-element)
+entry pointing to the right assemblies, because this requires knowing the exact assembly versions. 
+
+### Microsoft.DotNet.MSBuildSdkResolver
+
+This is the base resolver, capable of resolving "in-box" SDKs that ship with the .NET SDK, and workloads. Since the resolver assembly
+is located at a known path relative to MSBuild and has very few dependencies, none of which are used anywhere else, we have decided to
+freeze the version of the resolver plus dependencies, so that their full names can be specified in MSBuild.exe.config, e.g.
+
+```
+    <dependentAssembly>
+      <assemblyIdentity name="Microsoft.DotNet.MSBuildSdkResolver" culture="neutral" publicKeyToken="adb9793829ddae60" />
+      <codeBase version="8.0.100.0" href=".\SdkResolvers\Microsoft.DotNet.MSBuildSdkResolver\Microsoft.DotNet.MSBuildSdkResolver.dll" />
+    </dependentAssembly>
+```
+
+Additionally, MSBuild.exe.config has the following entry, which enables us to refer to the resolver by simple name.
+
+```
+<qualifyAssembly partialName="Microsoft.DotNet.MSBuildSdkResolver" fullName="Microsoft.DotNet.MSBuildSdkResolver, Version=8.0.100.0, Culture=neutral, PublicKeyToken=adb9793829ddae60" />
+```
+
+This has a small advantage compared to hardcoding `Microsoft.DotNet.MSBuildSdkResolver, Version=8.0.100.0, Culture=neutral, PublicKeyToken=adb9793829ddae60`
+directly in the code, as it can be modified to work in non-standard environments just by editing the app config appropriately.
+
+The resolver loading logic in MSBuild has been updated to call `Assembly.Load(AssemblyName)` where the `AssemblyName` specifies the
+simple name of the assembly, e.g. `Microsoft.DotNet.MSBuildSdkResolver`, as well as its `CodeBase` (file path). This way the CLR assembly
+loader will try to load the assembly into the default context first - a necessary condition for the native image to be used - and fall back
+to LoadFrom if the simple name wasn't resolved.
+
+### Microsoft.Build.NuGetSdkResolver
+
+The NuGet resolver has many dependencies and its version is frequently changing. Due to the way Visual Studio is composed, MSBuild does
+not know at its build time the exact version it will be loading at run-time. We would need a creative installer solution to be able to
+have MSbuild.exe.config contain the right entries to load this resolver the same way as `Microsoft.DotNet.MSBuildSdkResolver``. A more
+promising direction is loading this resolver into a separate AppDomain (see the section about NuGet.Frameworks below).
+
+The problem of JITting `Microsoft.Build.NuGetSdkResolver` remains unsolved for now.
```

Comment:
@jeffkl I believe you've also looked at optimizing NuGet resolver startup/JITting. Curious if you have any insights. Thank you!

Reply:
I was going to investigate doing the following:

1. ILMerge everything into the NuGet-based MSBuild project SDK resolver so its only a single assembly
2. Stop changing the assembly version and only change the informational version
3. Ship it next to MSBuild by "inserting" it into MSBuild or add a `codeBase` entry in the `MSBuild.exe.config` since it can be relative.

#1 provided a surprising amount of improvement but some people are opposed to merging 
#2 should be a no-brainer in my opinion, I stopped rev'ing assembly versions a long time ago for everything I own

I did a test of this in early 2023 where I ILMerged the SDK resolver, manually edited the `MSBuild.exe.config` and built a local copy of MSBuild with changes to SDK resolver loading that worked against an assembly name rather than a path.  I saw a big performance increase for startup times when using the NuGet-based MSBuild project SDK resolver.  Let me know if you want me to dig up the branch where I made changes to MSBuild to load SDK resolvers by name instead of path.

Reply:
Thank you for the prompt response.

> Let me know if you want me to dig up the branch where I made changes to MSBuild to load SDK resolvers by name instead of path.

This is actually [already merged](https://github.com/dotnet/msbuild/pull/9439) and shipping in 17.9, affecting only `Microsoft.DotNet.MSBuildSdkResolver` which we were able to version-freeze. The impact on startup perf was indeed quite dramatic and it would be awesome if we can apply to the NuGet resolver as well.

Is there anything I can do to help make this happen?

Reply:
The NuGet-based MSBuild project SDK resolver is loaded [via a manifest](https://github.com/dotnet/msbuild/blob/main/src/MSBuild/SdkResolvers/VS/Microsoft.Build.NuGetSdkResolver.xml), I had originally envisioned us updating the manifest reading code to look for a `name` element instead of a `path` but I suppose your change would work as well.  We'd need to ILMerge the SDK resolver and lock the version which is pretty simple.  Unfortunately, I'm not sure I'll have time in January to work on this.